File: server/src/models/module.rs
use diesel;
use diesel::prelude::*;

use uuid::Uuid;
use chrono::{DateTime, UTC};

use helpers::db::DB;
use schema::modules;
use schema::modules::dsl as module_dsl;


#[derive(Debug, Serialize, Deserialize, Identifiable, Queryable, Associations)]
#[belongs_to(pods)]
#[belongs_to(bases)]
pub struct Module {
    pub name: String,
    pub level: i32,
    pub stationary: bool,
    pub x_pos: Option<i32>,
    pub y_pos: Option<i32>,

    pub id: Uuid,
    pub pod_id: Option<Uuid>,
    pub base_id: Option<Uuid>,

    pub created_at: DateTime<UTC>,
    pub updated_at: DateTime<UTC>,
}


impl Module {
    /// This function adds a new pod into the database and returns the model with the
    /// initialized data.
    pub fn get(id: Uuid, db: &DB) -> Result<Module, diesel::result::Error> {
      // Get the research from an id
      module_dsl::modules
          .filter(module_dsl::id.eq(id))
          .get_result::<Module>(&**db)
    }
}


#[derive(Insertable)]
#[table_name="modules"]
pub struct NewModule {
    pub name: String,
    pub stationary: bool,
    pub x_pos: Option<i32>,
    pub y_pos: Option<i32>,
    pub pod_id: Option<Uuid>,
    pub base_id: Option<Uuid>,
}

File: server/models/module.py
"""The file for the Module database model."""

import uuid
from sqlalchemy.orm import relationship
from sqlalchemy import (
    func,
    Column,
    CheckConstraint,
    ForeignKey,
)
from sqlalchemy.types import (
    Boolean,
    String,
    Integer,
    DateTime,
)
from sqlalchemy.dialects.postgresql import UUID

from server.extensions import db


class Module(db.Model):
    """The database model for all Modules."""

    __tablename__ = 'module'
    __table_args__ = (
        CheckConstraint(
            "(pod_id is NULL and base_id is not NULL) or "
            "(pod_id is not NULL and base_id is NULL)"
        ),
    )

    id = Column(UUID(as_uuid=True), primary_key=True, default=uuid.uuid4)
    pod_id = Column(UUID(as_uuid=True), ForeignKey('pod.id'))
    base_id = Column(UUID(as_uuid=True), ForeignKey('base.id'))

    type = Column(String(255), nullable=False)
    level = Column(Integer, nullable=False)
    stationary = Column(Boolean, nullable=False)
    x_pos = Column(Integer)
    y_pos = Column(Integer)
    finished = Column(Boolean, nullable=False, default=False)

    pod = relationship("Pod", back_populates="modules")
    base = relationship("Base", back_populates="modules")

    created_at = Column(DateTime, server_default=func.now(), nullable=False)
    updated_at = Column(
        DateTime, server_default=func.now(),
        onupdate=func.current_timestamp(),
        nullable=False,
    )

    def __init__(self, type, pod, level, stationary, x_pos=None, y_pos=None):
        """Create a new module."""
        self.type = type
        self.pod = pod
        self.level = level
        self.stationary = stationary
        self.x_pos = x_pos
        self.y_pos = y_pos
